Ambiguous import "..." has multiple matching exports

### Describe the bug

I have a component (A) with a default export of the component and a named export in the same file. This component also contains an import.meta.glob and imports a directory (B).

Subsequently, Component A is imported into another component, C, which also contains an import.meta.glob of the same directory.

This results in:
```
Error:   Failed to scan for dependencies from entries:
  /home/projects/vue3-vite-starter-w2qd5j/index.html

  ✘ [ERROR] Ambiguous import "something" has multiple matching exports

    script:/home/projects/vue3-vite-starter-w2qd5j/src/App.vue?id=0:2:21:
      2 │ import HelloWorld, { something } from '@/components/HelloWorld.vue'
        ╵                      ~~~~~~~~~

  One matching export is here:

    script:/home/projects/vue3-vite-starter-w2qd5j/src/components/HelloWorld.vue?id=0:19:13:
      19 │ export const something = 'something'
         ╵              ~~~~~~~~~

  Another matching export is here:

    script:/home/projects/vue3-vite-starter-w2qd5j/src/components/HelloWorld.vue?id=1:19:13:
      19 │ export const something = 'something'
```

This issue arises only when the directory imported with import.meta.glob contains multiple files.

Overview:

![Untitled-2024-01-25-1600](https://github.com/vitejs/vite/assets/9058265/3e942971-4cc0-47cc-a262-fa80b54281c9)





### Reproduction

https://stackblitz.com/edit/vue3-vite-starter-w2qd5j

### Steps to reproduce

Run `yarn dev --force`

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.3 - /usr/local/bin/pnpm
  npmPackages:
    @vitejs/plugin-vue: ^5.0.4 => 5.0.4 
    vite: ^5.1.5 => 5.1.6
```


### Used Package Manager

yarn

### Logs

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/optimizer/scan.ts
+++ b/packages/vite/src/node/optimizer/scan.ts
@@ -409,12 +409,10 @@ function esbuildScanPlugin(
         // Avoid matching the content of the comment
         raw = raw.replace(commentRE, '<!---->')
         const isHtml = p.endsWith('.html')
-        scriptRE.lastIndex = 0
         let js = ''
         let scriptId = 0
-        let match: RegExpExecArray | null
-        while ((match = scriptRE.exec(raw))) {
-          const [, openTag, content] = match
+        const matches = raw.matchAll(scriptRE)
+        for (const [, openTag, content] of matches) {
           const typeMatch = openTag.match(typeRE)
           const type =
             typeMatch && (typeMatch[1] || typeMatch[2] || typeMatch[3])

